codama-koroks/src/impl_korok.rs

```diff
@@ -0,0 +1,48 @@
+use crate::{ConstKorok, KorokTrait, TraitKorok};
+use codama_attributes::Attributes;
+use codama_errors::{combine_errors, CodamaError, CodamaResult};
+use codama_nodes::Node;
+use codama_syn_helpers::extensions::*;
+
+#[derive(Debug, PartialEq)]
+pub struct ImplKorok<'a> {
+    pub ast: &'a syn::ItemImpl,
+    pub attributes: Attributes<'a>,
+    pub constants: Vec<ConstKorok<'a>>,
+    pub trait_: Option<TraitKorok>, // None means it's a Self impl
+    pub node: Option<Node>,
+}
```

Comment:
- Storing `items` makes it so our parsing doesn't loose any information by focusing on constants. And can just tap into the existing `ItemKorok` to recursively get the rest of the items (including constants). The only gotcha here is `ItemKorok` relies on `syn::Item` instead of `syn::ImplItem` so we might need to introduce an enum to allow both (or more in the future such as `syn::TraitItem`).
- Storing `trait_` here is redundant since the information is already on `ast`.

```suggestion
pub struct ImplKorok<'a> {
    pub ast: &'a syn::ItemImpl,
    pub attributes: Attributes<'a>,
    pub items: Vec<ItemKorok<'a>>,
    pub node: Option<Node>,
}
```

Reply:
> * Storing `items` makes it so our parsing doesn't loose any information by focusing on constants.

Actually this was my intention. I didn't want to have to write a korok for everything that can be included in an `impl` block. Is it going to be difficult to build on this in the future if we stop at constants now?

Fwiw, it sounds like we'd actually need an `ImplItemKorok`, to go along with the attribute context variant.

> * Storing `trait_` here is redundant since the information is already on `ast`.

Yeah it is a little bit, but I felt like having to always do this if you need to obtain the trait name kinda stunk:

```rust
impl_korok
    .ast
    .trait_.as_ref().map(|(_, path, _)| path)
    .segments
    .iter()
    .map(|seg| seg.ident.to_string())
    .collect::<Vec<_>>()
    .join("::")
```

This way, you get a nice little short-hand way to identify if it's one of these:

```rust
impl MyType { }
```

or these:

```rust
impl MyTrait for MyType { }
```

And if you need the fully-qualified path, or any other information about the trait, you can go through the `ast`.

Reply:
Hmm I still don't think that parsing is necessary because we have `syn` extensions for that sort of logic. For instance, [`path.idents()`](https://github.com/codama-idl/codama-rs/blob/main/codama-syn-helpers/src/extensions/path.rs#L8-L21) does most of that logic. All you need is unwrapping the path which doesn't warrant a new Rust item IMO.

Reply:
Alright @lorisleiva maybe you had a point here! I added the `ImplItemKorok` and a lot of the old API came out with the wash. Let me know what you think.